zebra-state/src/service/memory_state.rs

```diff
@@ -460,11 +402,199 @@ impl Ord for Chain {
     }
 }
 
+/// The state of the chains in memory, incuding queued blocks.
+#[derive(Default)]
+pub struct NonFinalizedState {
+    /// Verified, non-finalized chains.
+    chain_set: BTreeSet<Box<Chain>>,
+}
+
+impl NonFinalizedState {
+    /// Finalize the lowest height block in the non-finalized portion of the best
+    /// chain and update all side-chains to match.
+    pub fn finalize(&mut self) -> Arc<Block> {
+        let chains = mem::take(&mut self.chain_set);
+        let mut chains = chains.into_iter();
+
+        // extract best chain
+        let mut best_chain = chains.next().expect("there's at least one chain");
+        // extract the rest into side_chains so they can be mutated
+        let side_chains = chains;
+
+        // remove the lowest height block from the best_chain as finalized_block
+        let finalized_block = best_chain.pop_root();
+        // add best_chain back to `self.chain_set`
+        self.chain_set.insert(best_chain);
+
+        // for each remaining chain in side_chains
+        for mut chain in side_chains {
+            // remove the first block from `chain`
+            let chain_start = chain.pop_root();
+            // if block equals finalized_block
+            if chain_start == finalized_block {
+                // add the chain back to `self.chain_set`
+                self.chain_set.insert(chain);
+            }
+            // else discard `chain`
+        }
+
+        // return the finalized block
+        finalized_block
+    }
+
+    /// Commit block to the non-finalize state.
+    pub fn commit_block(&mut self, block: Arc<Block>) {
+        let parent_hash = block.header.previous_block_hash;
+
+        let mut parent_chain = self
+            .take_chain_if(|chain| chain.non_finalized_tip_hash() == parent_hash)
+            .or_else(|| {
+                self.chain_set
+                    .iter()
+                    .find_map(|chain| chain.fork(parent_hash))
+                    .map(Box::new)
+            })
+            .expect("commit_block is only called with blocks that are ready to be commited");
+
+        parent_chain.push(block);
+        self.chain_set.insert(parent_chain);
+    }
+
+    pub fn commit_new_chain(&mut self, block: Arc<Block>) {
+        let mut chain = Chain::default();
+        chain.push(block);
+        self.chain_set.insert(Box::new(chain));
+    }
+
+    pub fn best_chain_len(&self) -> block::Height {
+        block::Height(
+            self.chain_set
+                .iter()
+                .next()
```

Comment:
The best chain has the highest work.

Is there a test for the best chain in this function?
If there is, that test is also incorrect.
If not, please add one.

It might also be helpful to create a separate `best_chain` function.

```suggestion
                .next_back()
```

Reply:
Un-resolved, because this bug should also have a working test.

Reply:
Added the todo in the follow up pr for tests